source/Halibut/ServiceModel/PendingRequestQueue.cs

```diff
@@ -151,20 +162,42 @@ public PendingRequest(RequestMessage request, ILog log)
                 waiter = new ManualResetEventSlim(false);
             }
 
-            public RequestMessage Request
-            {
-                get { return request; }
-            }
+            public RequestMessage Request => request;
 
             public void WaitUntilComplete(CancellationToken cancellationToken)
             {
                 log.Write(EventType.MessageExchange, "Request {0} was queued", request);
 
-                var success = waiter.Wait(request.Destination.PollingRequestQueueTimeout, cancellationToken);
-                if (success)
+                bool success;
+                var cancelled = false;
+
+                try
                 {
-                    log.Write(EventType.MessageExchange, "Request {0} was collected by the polling endpoint", request);
-                    return;
+                    success = waiter.Wait(request.Destination.PollingRequestQueueTimeout, cancellationToken);
+                    if (success)
+                    {
+                        log.Write(EventType.MessageExchange, "Request {0} was collected by the polling endpoint", request);
+                        return;
+                    }
+                }
+                catch (Exception ex) when (ex is TaskCanceledException or OperationCanceledException)
+                {
+                    // waiter.Set is only called when the request has been collected and the response received.
+                    // It is possible that the transfer has already stated once the cancellationToken is cancelled
+                    // In this case we cannot walk away from the request as it is already in progress and no longer in the connecting phase
+                    cancelled = true;
+
+                    lock (sync)
+                    {
+                        if (!transferBegun)
```

Comment:
If we haven't begun transfer we are safe to cancel

Reply:
Minor typo in the comment "stated" should be "started"